### PR TITLE
[Merged by Bors] - generate MalfeasanceProof for ATXs and Ballots

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

1 of 2 PRs for #3919

## Changes

- create 3 types of MalfeasanceProof: ATX, ballots, and hare equivocation
- MalfeasanceProof gossip handler
- generate ATX MalfeasanceProof when receiving multiple ATX for the same epoch
- generate Ballot MalfeasanceProof when receiving multiple ballots for the same epoch

note: malicious ATX/proposals/ballots are all saved for now, pending research conclusions. only identities are marked as malicious.

### CODE REVIEW

File: common/types/malfeasance.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+type MalfeasanceProof struct {
+	// for network upgrade
+	Layer LayerID
+	Type  MalfeasanceType
+	// conflicting messages signed by the same NodeID
+	Messages []any
+}
+
+func (t *MalfeasanceProof) EncodeScale(enc *scale.Encoder) (int, error) {
```
Comment (countvonzero) [inline common/types/malfeasance.go]: @dshulyak please advise if improvements can be made in the custom encode/decode. thank you

**Comment by countvonzero** [comment]
bors try

File: common/types/malfeasance.go (hunk 1 above)
Comment (dshulyak) [reply common/types/malfeasance.go]: i would use a following approach:

1. define different malfeasence proofs. all of them can be autogenerated

```go
type AtxProof struct {
    [2]ConcreteAtxMessage
}

type HareProof struct {
   [2]ConcreteHareMessage
}
```

2. define enum for them. this is the only object that needs custom encoding, as the library doesn't support autogenerating it and it is not straighforward, but in general should be possible to autogenerate it with some work

```go

type ProofEnum struct {
    Type uint8
    Proof scale.Type
}

func (e *ProofEnum) EncodeScale(e *scale.Encode) {
    // not compact, as scale spec uses "full" uint8 for enums
    scale.EncodeByte(e, byte(e.Type))
   e.Proof.EncodeScale(e)
}

func (e *ProofEnum) DecodeScale(d *scale.Decoder) {
   typ := scale.DecodeByte(d)
   switch typ {
      // handle concrete type
   }
}
```  

3. higher level malfeasence proof for versioning. autogenerated

```go
type MalfeasenceProof struct {
    Layer types.LayerID
    Enum ProofEnum
}

```

**Comment by countvonzero** [comment]
bors try

File: common/types/malfeasance.go (hunk 1 above)
Comment (countvonzero) [reply common/types/malfeasance.go]: @dshulyak done as suggested with some name changes. thanks for helping.

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Review by dshulyak — APPROVED** [review]

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**